### ISSUE TITLE
Discussions

### ISSUE THREAD

**Issue by philipturner** [dl4s-team/dl4s-evolution#4]
@RahulBhalley, @palle-k did you receive alerts when I made the new post in the discussions section?

**Comment by philipturner** [comment]
@ryendu, @digantamisra98 did you get the alert I'm talking about? I'm just trying to troubleshoot something and an issue is the most effective way of doing it.

**Comment by ryendu** [comment]
> @ryendu, @digantamisra98 did you get the alert I'm talking about? I'm just trying to troubleshoot something and an issue is the most effective way of doing it.

nope i did not

**Comment by philipturner** [comment]
Okay. I’m going to post large updates in discussions from now on, then comment in this thread to send you all alerts. @palle-k, @RahulBhalley - please read what I have posted in discussions.

**Comment by digantamisra98** [comment]
> @ryendu, @digantamisra98 did you get the alert I'm talking about? I'm just trying to troubleshoot something and an issue is the most effective way of doing it.

I got it on email

**Comment by philipturner** [comment]
Finished work on single-precision matrix-vector multiply. Obliterated Metal Performance Shaders in edge cases, and doubled main memory throughput by using BFloat16 instead of Float32. However, the bandwidth savings make no difference (sometimes slightly slower) when the model falls into the M1 Max’s humongous system-level cache (64 MB), which happens quite a lot.

BFloat16 and sparse layers (an alternative to dense layers) were the two things I was keeping secret. I tried my hardest to skip vectorized memory reads with 4x4 (or higher) block sparsity to accelerate training, but it only harmed performance for BFloat16 (helped for Float32 - giving both data types almost the same performance). Sparsity will have no effect on main memory bandwidth, and you have to select BFloat16 as a layer type if you want the fastest training. However, after the model is converted to ONNX and used in inference, sparsity will improve model execution time. In addition, it also improves accuracy by acting sort of like a regularizer.

I am looking at the [SWAT](https://proceedings.neurips.cc/paper/2020/file/b44182379bf9fae976e6ae5996e13cd8-Paper.pdf) algorithm for sparsity, which improves on Google’s [RigL](https://arxiv.org/abs/1911.11134) and gives more flexibility for how neuron connections are rearranged during training. @digantamisra98 I’m interested in what you think about this, since you’re an established ML researcher.

Implementing batch-level parallelization will be trivial, so skipping it for now to work on convolutions. Plans to obliterate MPS in there as well, even use the new `simd_shuffle_and_fill` instruction on the iPhone 13’s A15 GPU. Batch-level parallelization can multiply by the depth of each convolution layer to give a massive amount of parallelization - fully utilizing the M1 Max’s 8192 32-bit ops/clock cycle. I originally thought this kind of parallelization would only happen post-im2col, then I did some actual research into CNNs and realized there are depthwise convolutions :). Going to use raw buffer memory instead of textures, as it’s much more flexible, low-overhead and I’m not using the texture sampling hardware. @ryendu I’m interested if you have any comments about this since you pushed so hard for im2col, or if you got lost in my talk about GPU optimization.

Also have more good news for the JIT compiler. The overhead seems so be incredibly low, on the order of microseconds. That means you can run an empty model _100 thousand_ times per second. The JIT compiler will also recognize when you abuse DL4S and loop over data in a batch instead of explicitly parallelizing it. This is similar to vectorization of loops in a CPU code compiler. However, it can’t execute infinitely many loop iterations per command buffer, as that would stall an Apple silicon Mac. I had to restart mine when I made a never-ending command buffer, because macOS doesn’t have the UI watchdog that iOS has.

Still keeping my progress in a personal private repo until it’s good enough to become a private repo of @dl4s-team.

Thanks for reading this!

**Comment by palle-k** [comment]
@philipturner I will sort through the rest of your input this week. 

Linux support is somewhat important, because if you want to train a model for longer, chances are you're gonna do it in a data center on a Linux machine. 

Also, some apps may use server side Swift and having the option to do server side inference there would be nice. 

So, if it isn't too big of an effort to maintain, Linux compatibility would be useful. 

Regarding watchOS: Of course there are tight constraints around memory and other resources but I could still imagine DL4S to be useful for applications such as handwriting recognition, swipe keyboard implementations, basic language modeling for predictive text, etc. 

DL4S was meant to be platform independent and running on every platform that Swift runs on.

**Comment by palle-k** [comment]
I imported the DL4S repository to this organization. I'll add deprecation notices and further information to my instance of DL4S soon, so that users are redirected towards dl4s-team/DL4S.

**Comment by philipturner** [comment]
I’d rather hold off on transferring DL4S to the organization right now. I’m about to make a long comment explaining why I’m considering dropping support for those platforms, but I’d rather tell you my thoughts about transferring as soon as possible.

**Comment by philipturner** [comment]
Please forgive me for the extremely long comment I have below. I encourage you to read this, even if you skip reading my updates above on GPU support.

The reason I thought about dropping those platforms isn't limitations to memory. It's that I would have to invest time in a decently fast CPU alternative to every GPU op added in DL4S 1.0 (e.g. Conv3D). I’m an expert in GPU optimization, not CPU optimization - look at someone like @michaeleisel for that. Although DL4S 1.0 could run on a CPU-only platform, it will be so slow that I'd rather not do that at all.

Say, for example, an added atomic operation (e.g. Conv3D) performs 10 GFLOPS at the theoretically fastest optimization level on the CPU. On CPU code without any care put into it, it would run at 0.1 GFLOPS. But on GPU, it would outperform MLCompute or Metal-accelerated TensorFlow and reach 100 GFLOPS or several TFLOPS. This is more than a 1000x difference, making the CPU use case pointless. I don't want to have such a disparity and an option for people to waste so much CO2 with a severely under-optimized CPU mode.

> Regarding watchOS: Of course there are tight constraints around memory and other resources but I could still imagine DL4S to be useful for applications such as handwriting recognition, swipe keyboard implementations, basic language modeling for predictive text, etc.

DL4S is not an inference framework. It's for the process of training, and only that. CoreML models are much more appropriate for inference than firing up DL4S's future JIT compiler and command buffer pipeline (which has an insane overhead for inference use cases). We will export to ONNX, which can then be converted to an optimized MLModel for on-device inference on an Apple Watch, or to TFLite for Android and edge devices.

> Linux support is somewhat important, because if you want to train a model for longer, chances are you're gonna do it in a data center on a Linux machine.

Your code for convolutions appears terribly un-optimized. I haven't 100% checked it, but it doesn't have the assembly-level optimization and cache optimization as TensorFlow or PyTorch's C++ code. Being so slow, it would be more appropriate for someone to use S4TF, or interact with TensorFlow or PyTorch through PythonKit.

> to train a model for longer

For that use case, it's much more economical to run TensorFlow in a Colab notebook on a cloud GPU or TPU, rather than what we have now. CPU is just a lot slower than GPU. It will take orders of magnitude less time to train it on your Mac or iPhone than on CPU on a server. For example, it might take days on the server which can't have GPU acceleration, but take less than an hour on an Intel Mac mini's integrated GPU.

I’m all for bringing it to other platforms, but I do not see that as a reasonable use case for this framework. We need to import a CPU framework that does more than just matrix multiplications - it needs fully optimized convolutions. MKL isn’t an option for that. We will need to import CPU TensorFlow or PyTorch through PythonKit, and I don’t anticipate being the person to invest time doing that. If you find a decent Linux equivalent to BNNS, maybe that will work.

Another alternative is bringing this to Vulkan, which would allow for Windows support, although server-side Linux usually does't let you access the GPU except in a Colab Notebook. I don't anticipate having the time or motivation to do anything other than Metal. I already anticipate having to spend 1000 hours to bring Metal support, and that's the easiest of all the GPGPU frameworks. The massive difference in ease of use between Metal and Vulkan is one reason why I recently decided to drop my plans for bringing ARHeadsetKit to Android.

I haven’t profiled DL4S on certain ops, and correct me if you have solid evidence there's no performance difference with third party frameworks. I'm mostly worried about the new ops and the added complexity of redirecting API calls to the JIT compiler to alternative code on CPU-only platforms.

I was considering directing people to your old version of DL4S if they wanted to use server-side Swift. The old version wouldn't have the new ops or API. That thought is why I was hesitant about you being too strong about deprecating the current DL4S.

-----

At your request, I will keep support for watchOS, Linux, and add Windows, but expect it to be insanely slow on those platforms.

**Comment by philipturner** [comment]
Also, it's awkward to have a BFloat16 tensor or layer type that doesn't actually operate on BFloat16 in CPU mode.

**Comment by philipturner** [comment]
I take back a little of what I said. Supporting Vulkan isn’t impossible, and a lot of the time on the Metal implementation is me doing research into ML. Design concepts from the Metal implementation should be repurposed able into Vulkan.

This will also solve the cross-platform issue, since people on Windows PCs will be able to train a model on their personal Windows PC’s GPU several times faster than on a server CPU.

**Comment by philipturner** [comment]
We could also make a better version of SwiftCoreMLTools, in addition to the option to export to ONNX, that lets this be used for inference on an Apple watch. We would just discourage or disable training on an Apple watch, since that’s so ridiculously slow.

You could just offload the training computation to the paired iPhone instead of wasting one of the watch’s two CPU cores.

**Comment by philipturner** [comment]
How about this plan: I bring it to macOS, iOS, and tvOS on the initial release of DL4S 1.0. You tell others to use the old DL4S if they don’t have an Apple device.

For an Apple Watch, DL4S requires a connection to an iPhone to do any training. API calls could still be done on the watch’s CPU, but they would transfer to the iPhone over bluetooth the moment training starts. The iPhone would finish orders of magnitude faster than the watch ever could, sending training results back after it finishes, or when any requests to a tensor’s raw contents interrupt execution.

Then, I bring it to Vulkan. Once that is working, open up Linux and Windows for machines that give access to GPU. It could also run on Android, if Swift ever supports that in the future.

**Comment by philipturner** [comment]
Also, dynamic device mode will not happen, and double precision support will be dropped in DL4S 1.0.

**Comment by digantamisra98** [comment]
@philipturner I will read the SWAT paper this week and get back to you.

**Comment by philipturner** [comment]
We may have a use for @differentiable - for constructing custom layers. Complex models like GoogleLeNet and ResNet have layers which don't follow the `Sequential` API of Keras or DL4S 0.x. However, I can't think of a way to get a `Concurrent` equivalent to work.

I think that for a case where you need to bind layers in custom ways (e.g. Inception modules) or propagate gradients in custom ways (e.g. GoogLeNet's 3 entry points for loss gradients), there should be a S4TF-like API. You could build modules of custom layers, which could then fit into the current Sequential API. 

The JIT compiler would call the custom layer's `callAsFunction()` and the automatically synthesized gradient once, generating a sequence of IR instructions. One could also configure the entire model manually and without the `Sequential` API, as long as it received the same IR.

**Comment by philipturner** [comment]
I'm also going to remove higher-order gradients from DL4S 1.0. There are very few reasonable use cases for them, yet they would tremendously complicate a JIT compiler. I have seen literature presenting an efficient way to propagate second-order gradients through fully connected layers. I can't say the same for convolutions, and definitely not for bilinear resizing filters, which require synchronization through atomic memory instructions.

I plan to make a new `Gradient` type in the public API, and which results from applying a gradient to a model. There will no longer be an option to specify `retainBackwardsGraph` when initializing a tensor. @palle-k, what are your feelings on this?

**Comment by philipturner** [comment]
Just finished work on random number generation. The GPU can make an arbitrarily long tape of bits, then consume it to produce uniform/normal/bernoulli distributed numbers, or use it for dropout layers or rearrangement of connections in a sparse layer. This is measurably faster than Apple's MPS kernels, and more flexible since MPS can't be used to produce random integers.

**Comment by palle-k** [comment]
Regarding 2nd order gradients: There are definitely use cases out there, like enforcement of Lipschitz constraints in WGAN-GP and similar models and by composing the backwards pass out of forward pass operations, support for higher order gradients is achievable. 

I would be fine with not every operation being differentiable multiple times, only the core of operations that are being used within neural networks. Preprocesing stuff like image resizing and so on doesn't need to be differentiable an arbitrary number of times. 

This feature is actually one of the features that is close to my heart because I've done some research on WGANs in the past. If you say that it's not easily achievable, then okay.

**Comment by palle-k** [comment]
I implemented higher order gradients by composing the backwards pass out of operations that are differentiable. 

Then I can just backpropagate through the compute graph of the forward and backward pass and that's it. There is not much more to it. 

Of course, the one big difference is that with DL4S' previous focus on CPU computation, I was able to not use fused ops, so this was pretty straightforward to implement. 

One organizational thing: Let's untangle discussions about different topics into dedicated issues, which are easier to follow.

**Comment by philipturner** [comment]
> One organizational thing: Let's untangle discussions about different topics into dedicated issues, which are easier to follow.

Good idea. I just had a massive change of plans that will speed up the implementation of GPU support. I'm going to close off this thread with that, then make separate issues for new topics I'm considering.

My attempt to make matrix multiplication and convolution shaders that outperformed Metal Performance Shaders didn't work out, and I'm going to rely on MPS a lot more than I thought I would. Also, I realized that MPS does have batch-level parallelization. In fact, it's implemented by every major ML API, but I wasn't aware of that when I started exploring.

> I implemented higher order gradients by composing the backwards pass out of operations that are differentiable.

Being able to compose higher-order gradients out of already existing functions means I can rely on MPS shaders for that. This is good news.

I found a new solution that lets me hide the latency of encoding MPS's shaders, so I don't need to make custom shaders just for the purpose of reducing encoding time. The solution is ahead-of-time compilation, and there won't be a big need for ICBs anymore. This is also a good idea for Vulkan, which doesn't have ICBs.

> Of course, the one big difference is that with DL4S' previous focus on CPU computation, I was able to not use fused ops, so this was pretty straightforward to implement.

A GPU engine lacking support for fused ops isn't really a bottleneck (which is good news). As I've discovered the hard way, it boils down to 3 factors:
- duration of matrix multiplications
- duration of convolutions
- encoding overhead

Not using custom shaders for the most performance-intensive ops means dropping support for `BFloat16` and sparsely connected layers. I think the overhead of both of those features would actually backfire and harm performance anyway. I'm still going to make several custom shaders for things like the Mish activation function, random number generation, and im2col.

@digantamisra98 you don't need to read the research paper on sparse neural networks now.

One last thing. I don't think my JIT compiler will auto-vectorize loops. If you execute everything in a batch serially in your Swift code, you'll get the consequences of that.
